### PR TITLE
Adding VRF and VXLAN support for systemd-networkd (LP: #1764716, LP: #1773522)

### PR DESCRIPTION
**Pull request by netdever**
## Description

Adding VRF and VXLAN support for systemd-networkd.

## Checklist

- [x] Runs `make check` successfully.
- [ ] Retains 100% code coverage (`make check-coverage`).
- [x] New/changed keys in YAML format are documented.
- [x] \(Optional\) Adds example YAML for new feature.
- [x] \(Optional\) Closes an open bug in Launchpad. [LP#1764716](https://pad.lv/1764716), [LP#1773522](https://pad.lv/1773522)

### CODE REVIEW

File: src/parse-nm.c
Comment (Conan-Kudo) [inline src/parse-nm.c:65]: I'm pretty sure this is supported in the keyfile format these days... @thom311, @lkundrak, @tyll, do any of you know?

**Review by Conan-Kudo — CHANGES_REQUESTED** [review]

File: src/parse-nm.c
Comment (thom311) [reply src/parse-nm.c:65]: NetworkManager's keyfile format supports (necessarily) everything that NetworkManager supports.

But I think the reference to "the keyfile parser" here is `src/parse-nm.c` from netplan itself. This branch doesn't seem to implement the NetworkManager backend (which is a fine choice).

For the record, NetworkManager supports vxlan and vrfs.

Comment (netdever) [reply src/parse-nm.c:65]: That's correct - I intentionally did not touch the nm parsers in this pr. Thanks!

Comment (Conan-Kudo) [reply src/parse-nm.c:65]: @thom311 Is there any documentation of this stuff on networkmanager.dev? I couldn't find it specifically.

@netdever While I'm not part of the project, I would personally prefer to see things like this being implemented for both backends.

Comment (thom311) [reply src/parse-nm.c:65]: in NM, you always create "connection profiles". So, to create a vxlan/vrf configuration, you also create appropriate profiles.  The fields of a profile are documented in `man nm-settings-{dbus,nmcli,keyfile}` (which is also on the website). But there is no direct documentation how to create a vxlan/vrf configuration.

File: src/abi.h
Comment (slyon) [inline src/abi.h:331]:
```suggestion
    /* vxlan */
    /* netplan-feature: vxlan */
```

Please let's add a feature flag here for the new VXLAN feature. Also, maybe move the VXLAN fields below the new VRF fields, so they are listed together with the `vxlan_params` and not intertwined.

File: src/abi.h
Comment (slyon) [inline src/abi.h:338]:
```suggestion
    /* vrf */
    /* netplan-feature: vrf */
```

Let's add a feature-flag for VRF, too

File: src/abi.h
Comment (slyon) [inline src/abi.h:368]: Thank you for moving any new fields in this `netplan_net_definition` struct to the very end, this way we'll only append the ABI and not break it :-)

File: src/netplan.c
Comment (slyon) [inline src/netplan.c:256]: In which way would we specify a "port-range" value? Can you give an example for this?

It seems like NetworkManager uses two separate settings for this:
* source-port-max
* source-port-min

Combining those into a single "port-range" sounds nice, but I'm wondering about the syntax?

https://networkmanager.dev/docs/api/latest/nm-settings-nmcli.html#id-1.2.9.4.43

File: src/netplan.c
Comment (slyon) [inline src/netplan.c:238]: IIUC Networkmanager uses a single setting to specify "remote" and "group", I wonder if we could do the same? Would the "remote" and "group" settings ever be used at the same time? Or can we deduct from the setting's value which one was meant by the user?

https://networkmanager.dev/docs/api/latest/nm-settings-nmcli.html#id-1.2.9.4.43

File: src/parse-nm.c
Comment (slyon) [reply src/parse-nm.c:65]: FTR: NetworkManager's VRF and VXLAN settings are listed here: https://networkmanager.dev/docs/api/latest/nm-settings-nmcli.html#id-1.2.9.4.43

But as Thomas stated, this code snippet in `src/parse-nm.c` is from netplan itself (allowing it to read a NM keyfile connection profile) and not necessarily needed for this PR.

**Review by slyon — COMMENTED** [review]
Hi @netdever
Thank you very much for your contribution to netplan, this is highly appreciated, especially as the two LP bugs that you're addressing have been long-standing feature requests!

I did an initial review and mentioned some smaller issues and questions in inline comments.

I am now also thinking about bigger changes that might be needed:
* configmanager.py to go away (conflict with #255)
* NetplanTristate type should be added (conflict with #270)
* Linking of network interfaces: I wonder if we should rather define a VRF's child interfaces inside the vrf definition itself, using an `interfaces: [...]` setting (as suggested in the LP#1773522) instead of splitting the definition across parent and child, using the `vrf: ...` link:
```
network:
  version: 2
  renderer: networkd
  vrfs:
    vrf-test:
      interfaces:
        - bond0
        - br0
        - br-bond0
        - tun0
      table: auto # or a specific integer
```
* Similarly, I need to thinkg about the linking of VXLAN devices, as the `ethX.vxlans.names = [...]` seems to be a bit cumbersome. Is there a specific reason why you went for this nestring structure?
* Overall, I want to review all the new YAML settings in detail, to make sure we implement a simple and easily comprehensible wording. There is nothing to change about this right now... for the current stage we can just go with systemd-networkd's naming scheme and we can easily change/adopt that, once agreed upon (but before we merge the PR :))
* NetworkManager also supports (most) VRF/VXLAN settings that are implemented in this PR for the systemd-networkd backend renderer. With all the parsing code already in place, it should be relatively easy to implement those in `src/nm.c`. Would you be interested in doing that? But it could also be done as part of a separate PR, in which case we should at least log a warning from the NetworkManager backend renderer, that those settings are not currently supported on that backend.
* Finally, I'd really love to see some actuall integration tests (in `tests/integration/...`) that make use of those new VRF and VXLAN features and check the fully configured system at runtime, to make sure the settings are actually applied. Is that something you could add do your PR? The tests are being run as `autopkgtests`, as can be seen (and reproduced locally) in ".github/workflows/autopkgtest.yml`.

File: src/abi.h
Comment (netdever) [reply src/abi.h:369]: No problem!

File: src/abi.h
Comment (netdever) [reply src/abi.h:337]: Done.

File: src/abi.h
Comment (netdever) [reply src/abi.h]: Done.

File: doc/netplan.md
Comment (netdever) [reply doc/netplan.md]: Done.

File: doc/netplan.md
Comment (netdever) [reply doc/netplan.md:1461]: I gave this a shot but hit some snags. I'm not sure optional-addresses behaves the same way, but I could be missing something.

File: netplan/configmanager.py
Comment (netdever) [reply netplan/configmanager.py:122]: Fingers crossed this gets merged first :)

File: src/abi.h
Comment (netdever) [reply src/abi.h:362]: Had some trouble trying to bundle these myself, so leaving as is for now. Any help would be appreciated if we feel it's necessary to make this change.

File: src/netplan.c
Comment (netdever) [reply src/netplan.c:256]: In my testing, I could only specify an integer for `PortRange=` using systemd-networkd. Obviously that's not a "range." Documentation is lacking for this parameter in systemd-networkd.

File: src/abi.h
Comment (netdever) [reply src/abi.h]: Good call, I overlooked that. I've updated the PR to use the existing definition.

File: src/netplan.c
Comment (slyon) [reply src/netplan.c:238]: Ineed, Group= will always be a multicast IP address (i.e. from a reserved IPv4/6 range), so we can implicitly derive this information, therefore I'd like to reduce those two fields to just a single `remote` field, that is than translated to Remote= or Group= for systemd-networkd depending on if it is a multicast IP or not.

File: doc/netplan.md
Comment (slyon) [inline doc/netplan.md:1483]: I'd like to avoid this setting, as it is very backend (systemd-networkd) specific. We already have this information if NetplanNetDefinition->vxlan_link is set and can set the value implicitly. No need to expose it to the user via YAML

Comment (netdever) [reply doc/netplan.md]: Sure - I removed this from the PR.

File: src/netplan.c
Comment (netdever) [reply src/netplan.c]: OK, I was able to get this to work as described. There isn't validation for it yet, though.

**Comment by netdever** [comment]
- Linking of VRFs now looks more like the example in the linked LP bug.
- Linking of VXLANs to source interfaces no longer requires the `names` subvar.
- VXLAN `source-port-range` now works as suggested (lacks validation)

File: src/abi.h
Comment (slyon) [inline src/abi.h:340]: This seems to be a rather generic `[Bridge]` setting, not necessarily related to VXLANs.. Is it required for VXLAN? If so, we should implement it as a generic bridge feature (and probably use a tristate type, too).

Otherwise, I'd like to just drop it from this PR, as it seems unrelated. Especially, as there is currently no implementation in any backend (neither `networkd.c` nor `nm.c` for this `neigh_suppress` variable).

Comment (netdever) [reply src/abi.h:340]: Neighbor suppression (arp-nd-suppression) is required for VXLAN configuration. It is disabled by default, but many implementations, specifically those utilizing EVPN, will need it to be enabled. In systemd-networkd, it looks like this:

```bash
root@host1:~# cat /run/systemd/network/10-netplan-vxlan10.network 
[Match]
Name=vxlan10

[Link]
MTUBytes=8950

[Network]
LinkLocalAddressing=no
IPv6AcceptRA=no
ConfigureWithoutCarrier=yes
Bridge=bridge10

[Bridge]
NeighborSuppression=True
```

Comment (netdever) [reply src/abi.h:340]: The command to enable this feature is ```bridge link set dev vxlan10 neigh_suppress on```

Comment (slyon) [reply src/abi.h:340]: Thank you for your comments about this. Being a `[Bridge]` setting in sd-networkd, I still wonder if this is a setting that should be set on the VXLAN device itself, or rather on a bridge device that is a parent interface to the vxlan?

Comment (netdever) [reply src/abi.h:340]: It is a VXLAN setting (ie. you cannot enable it on the master Bridge interface). It may be confusing for netplan to have it as a bridge setting, especially since enslaving a VXLAN to a Bridge is not technically a requirement.

```bash
root@host1:~# bridge link set dev bridge10 neigh_suppress on
RTNETLINK answers: Operation not supported
```

**Comment by slyon** [comment]
This PR got closed by accident after rebasing on top of the current `main` branch and then pushing the wrong branch to `netdever/main`. I do not have write access to netdever's branch anymore. So I created a copy of this in #285

**Pull request closed without merging**
